### PR TITLE
feat(anomaly): attribute auth-failure bursts to a client address (valkey#334)

### LINKED ISSUE #385: Advisory: authentication-failure / brute-force detection from ACL LOG client addresses (valkey#334)
## Summary

Add an **authentication-failure / brute-force advisory** built from `ACL LOG`, surfacing repeated auth failures **with the offending client address**. Operators want to know when a specific client IP is hammering the server with bad credentials or hitting denied commands/keys, not just that the aggregate `acl_access_denied_auth` counter ticked up. This maps to upstream valkey-io/valkey#334.

## Upstream context

- valkey-io/valkey#334 "[NEW] Log authentication failures with client address".
- Ask: security auditing of auth failures keyed by client IP. Upstream discussion points at `ACL LOG` (which already records client address, reason, and count) as the vehicle, plus possible future metric subcommands and an audit-log file - none of the richer additions have shipped, and the feature is deferred/undecided. Durable.
- We do not need the unshipped extras: `ACL LOG` already carries what we need today.

## What we can observe

We already have the plumbing - `getAclLog()` / `MetricsParser.parseAclLog`, the `hasAclLog` / `canAclLog` capability gates, the `/metrics/acl/log` API, and the `AclLogEntry` type - plus an existing `AUTH_ATTACK` correlator pattern and the `ACL_DENIED` metric (`acl_access_denied_auth`):

- `ACL LOG` entries: `reason` (auth / command / key / channel), `client-info` / `object`, `age-seconds`, and crucially the client `address`, with a `count` per aggregated entry.
- `ACL_DENIED` metric for the aggregate rate, to corroborate a spike.

## Detection logic

Each poll, read `ACL LOG` (bounded count) and aggregate by client address / username:
1. Flag when a single client address accumulates auth-failure (`reason=auth`) entries above a threshold within a window - a credential brute-force signature.
2. Optionally extend to `reason=command`/`key`/`channel` denials from one client as a misconfigured-or-probing-client advisory (a legitimate app suddenly hitting denials often means a bad deploy or ACL drift; a stranger hitting them means probing).
3. Corroborate with a rising `ACL_DENIED` (`acl_access_denied_auth`) rate so a single stale log entry does not fire.

Report the offending client address(es), username, failure count, and reason breakdown. Advise the operator to investigate the source IP, rotate credentials, or tighten `bind`/firewalling. Feed the finding into the existing `AUTH_ATTACK` pattern where appropriate.

## Scope and guardrails

- Capability-gated: only run when `hasAclLog` / `canAclLog` is true; degrade silently otherwise (no hard failure), matching the existing ACL-log capability checks.
- `ACL LOG` is a bounded ring (`acllog-max-len`) and can be wiped (`ACL LOG RESET`) or evicted - treat it as best-effort; use `count` and entry age, and do not assume completeness. De-duplicate across polls so the same aggregated entry is not counted repeatedly.
- Redact/limit sensitive material: report client address, username, reason, and counts - do not echo full command arguments or key names into the anomaly store / webhooks.
- Complementary to, not a duplicate of, the raw `ACL_DENIED` metric and the `AUTH_ATTACK` correlator: this advisory adds the per-client-address attribution those lack.

## Acceptance criteria

- New advisory under `proprietary/anomaly-detection/` with a dedicated `MetricType` (e.g. `AUTH_FAILURE_BURST`) in `types.ts` + `METRICS_HANDLED_OUTSIDE_EXTRACTOR` (ACL LOG is fetched outside the INFO extractor), wired into `anomaly.service.ts` pollConnection, with per-connection dedup state cleaned up in `onConnectionRemoved`.
- Unit tests: one client address over the auth-failure threshold in a window (fires, names the address), scattered single failures under threshold (silent), ACL-log capability absent (no-op), repeated polls of the same aggregated entry (counted once), ACL LOG RESET mid-window (no crash / no negative counts).
- Emitted event carries client address(es), username, failure count, and reason breakdown - no full argv/key material.
- UI label/description for the new metric type in `apps/web`.

### PR DESCRIPTION
**Pull request by jamby77**
Closes #385. **Stacked on #389** (→ #388 → #387) — this PR's own diff is the last
commit.

Adds an `AUTH_FAILURE_BURST` advisory: repeated failed AUTHs from **one client
IP**, which the aggregate `acl_access_denied_auth` counter cannot attribute.
Upstream #334 asks for exactly this; the richer additions discussed there (metric
subcommands, an audit-log file) never shipped and the issue is deferred, but
`ACL LOG` already carries reason, username, client-info and a per-entry count.

## It reads the audit store instead of polling ACL LOG again

`AuditService` already polls `ACL LOG` every cycle, gates on `hasAclLog` and the
`canAclLog` runtime tracker, dedupes, and persists `StoredAclEntry` rows. Polling
it a second time from the anomaly service would duplicate the fetch *and* the
dedup state.

Reading the store means the advisory inherits the capability gating, the
ring-buffer/`ACL LOG RESET` tolerance, and the cross-poll dedup already built
there. It also drops two acceptance criteria from the issue as no longer
applicable — "repeated polls of the same aggregated entry counted once" and
"ACL LOG RESET mid-window" — because the audit layer owns both.

Scan is throttled to 30s (the poll loop runs at 1s by default; the window is 5
minutes wide) and the query passes an explicit 2000-row limit — the store's
default is 100, which would silently truncate the window exactly when the
advisory matters most.

## Two traps this would have fallen into

**Cumulative counts.** An `ACL LOG` entry is an aggregate: repeated failures bump
one entry's `count` and its `timestamp-last-updated`. The audit poller stores a
NEW row whenever that timestamp advances, so one logical entry lands several
times with a *growing cumulative* count. Summing rows multiplies the real figure
— 3 + 7 + 15 reads as 25 failures when there were 15. Rows are collapsed to their
maximum per logical entry (keyed on `timestampCreated` + username + client-info +
reason) before aggregation.

**Ephemeral source ports.** Every connection attempt arrives from a different
source port, so `addr=1.2.3.4:53124` is unique per attempt. Aggregating on the
raw `addr` would put every single failure in its own bucket and never cross any
threshold — the detector would have been silently dead. The port is stripped and
only the client IP is the key.

Also worth flagging for the plan's sake: `StoredAclEntry.sourceHost`/`sourcePort`
are the **monitored server's** address, not the client's. The client address only
exists inside the `client-info` string and has to be parsed out of `addr=`.

## Redaction

The event carries the client address, usernames, counts and reason breakdown —
never the `object` field (key names, command names) or the raw `client-info`.
Those would land in the anomaly store and go out through webhook payloads. There
is an explicit test asserting a key name like `secret:customer:pii` cannot appear
in a message.

## Scope note

The issue also suggests corroborating with a rising `ACL_DENIED` rate "so a single
stale log entry does not fire". That is unnecessary here, but NOT for the reason
this section originally gave. The query is deliberately **not** windowed on
`captured_at` — that column is refreshed on every upsert, so it records when the
poller last saw the entry, not when the failures happened, and filtering on it
would select on poller behaviour. A stale entry is excluded because the detector
accumulates per-scan **growth**: an entry that is not climbing contributes zero,
however old or recent its row is. Non-auth
denials (`command`/`key`/`channel`) from the same address are still reported as a
breakdown — a legitimate app hitting them usually means a bad deploy, a stranger
hitting them means probing.

## Tests

- 16 unit tests: port stripping, IPv6 unwrapping, `laddr=` not mistaken for
  `addr=`, threshold, cumulative-row collapsing, distinct entries kept separate,
  brute force across 12 ephemeral ports, auth-only thresholding with full reason
  breakdown, worst-offender ranking, unparseable addresses skipped, `NaN` count,
  cooldown behaviour and map pruning.
- 8 service-level tests: event content, redaction, seconds-not-milliseconds
  window, 30s scan throttle, silence on an empty store, no re-alert on the next
  scan, storage failure not breaking the poll, state cleared on connection removal.
- 629/629 across all 26 anomaly suites; 311/311 web; `tsc --noEmit` clean.

Note: `getAclEntries` was missing from the service spec's storage mock, so it is
added here — without it every existing test would have exercised the new code's
error path.

---

> [!NOTE]
> **Medium Risk**
> New security-advisory path names client IPs operators may block; logic is complex (window growth, ACL LOG grouping limits) but heavily tested and failures are swallowed without breaking polls.
> 
> **Overview**
> Introduces an **`AUTH_FAILURE_BURST`** state-based advisory that attributes repeated failed AUTH attempts to a **single client IP**, using the existing ACL audit store instead of polling `ACL LOG` again.
> 
> A new **`auth-failure-detector`** module parses `client-info`, aggregates by IP (stripping ephemeral ports), tracks per-entry count **growth** over a 5-minute window (so lifetime totals and poller backfill do not false-positive), applies a per-address alert cooldown, and emits redacted events (no key names or raw `client-info`). **`AnomalyService`** runs a throttled (~30s) `getAclEntries` scan (2000-row limit), wires connection teardown cleanup, and surfaces the metric on the anomaly dashboard.
> 
> **Storage adapters** (memory, SQLite, Postgres) now update **`client_info` on ACL audit upsert**, so attribution follows the latest client seen on a logical entry; tests cover memory and SQLite.
> 
> **`MetricType.AUTH_FAILURE_BURST`** is registered for direct (non z-score) emission; unit and service tests cover detection edge cases, redaction, throttling, and resilience.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 76f8b238893876b423af2ebaed23143b076f3c90. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added detection for bursts of authentication failures from client addresses.
  * Alerts include failure counts and limited attribution details while avoiding sensitive data exposure.
  * Added “Auth Failure Burst” to the anomaly dashboard.
* **Bug Fixes**
  * ACL audit updates now refresh stored client information across memory, SQLite, and PostgreSQL storage.
  * Improved handling of repeated, stale, and malformed authentication-failure activity.

### CODE REVIEW

File: proprietary/anomaly-detection/anomaly.service.ts [hunk 1]
```diff
@@ -1402,6 +1424,117 @@ export class AnomalyService extends MultiConnectionPoller implements OnModuleIni
...
+    this.authFailureLastScan.set(ctx.connectionId, timestamp);
+
+    try {
+      // The audit store windows on `captured_at`, which is stored in SECONDS.
+      const windowStart = Math.floor((timestamp - AUTH_FAILURE_WINDOW_MS) / 1000);
+      const entries = await this.storage.getAclEntries({
+        connectionId: ctx.connectionId,
+        startTime: windowStart,
+        limit: AUTH_FAILURE_QUERY_LIMIT,
+      });
```
[the remark this answers is not in the document]
Comment (jamby77) [reply proprietary/anomaly-detection/anomaly.service.ts:1476]: Correct. Fixed across 09549b1d and a69f2675 — the second commit supersedes the first after your other finding showed the store upserts.

The query no longer filters on `captured_at` at all (an entry under active attack keeps its original `captured_at` while its count climbs, so filtering on it hid exactly the rows that matter). In-window activity is now measured as observed growth, and an entry seen for the first time only sets a baseline unless it was created inside the window — which is precisely the restart-backfill case. Test: `baselines an unseen entry rather than alerting on its lifetime count`.

File: proprietary/anomaly-detection/auth-failure-detector.ts [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    const count = Number.isFinite(entry.count) ? entry.count : 0;
+    bucket.reasonBreakdown[entry.reason] = (bucket.reasonBreakdown[entry.reason] ?? 0) + count;
+    if (entry.reason === AUTH_REASON) {
+      bucket.authFailures += count;
+      bucket.usernames.set(entry.username, (bucket.usernames.get(entry.username) ?? 0) + count);
+    }
+    bucket.firstSeenAt = Math.min(bucket.firstSeenAt, entry.capturedAt);
+    bucket.lastSeenAt = Math.max(bucket.lastSeenAt, entry.capturedAt);
+
+    byAddress.set(address, bucket);
```
[the remark this answers is not in the document]
Comment (jamby77) [reply proprietary/anomaly-detection/auth-failure-detector.ts:277]: Accurate, and not something I can engineer away — `ACLLogMatchEntry` matches on reason/context/object/username and never compares the client address, so one entry's count can span clients while `client-info` reflects only the most recent.

Rather than imply precision the data does not support, the claim is now weaker and explicit. The message reads "N authentication failures **recorded against** <addr>" and states that ACL LOG groups by user and reason rather than by client, so the address is the client recorded on those failures. The limitation is documented in the module header under "A limit worth knowing".

It remains strictly better than the aggregate `acl_access_denied_auth` counter, which offers no attribution at all — but it is a hint, not proof, and now says so.

File: proprietary/anomaly-detection/auth-failure-detector.ts [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    const counts = active.map((entry) => {
+      return Number.isFinite(entry.count) ? entry.count : 0;
+    });
+    const maxCount = Math.max(...counts);
+    const latest = active.reduce((newest, entry) => {
+      return entry.timestampLastUpdated >= newest.timestampLastUpdated ? entry : newest;
+    });
+
+    const bornInWindow = latest.timestampCreated >= windowStartMs;
+    const windowCount = bornInWindow ? maxCount : maxCount - Math.min(...counts);
```
[the remark this answers is not in the document]
Comment (jamby77) [reply proprietary/anomaly-detection/auth-failure-detector.ts]: This one was the most useful of the batch — thank you. Confirmed by reading the adapter: `ON CONFLICT(timestamp_created, username, object, reason, source_host, source_port, connection_id) DO UPDATE`, so one logical entry is exactly one row with its count rewritten in place. There is no row history to difference, and my `maxCount - minCount` was therefore always zero for anything created before the window.

Fixed in a69f2675 by tracking each entry's count between scans and accumulating per-scan growth across the window — the same treatment every other cumulative counter in this service gets. That also fixes a subtler version of the same problem you did not mention: a single per-scan delta would have split a sustained attack into 30-second chunks, each below threshold. Test: `keeps accumulating a sustained attack across several scans`.

File: proprietary/anomaly-detection/anomaly.service.ts [hunk 4]
```diff
@@ -1406,6 +1428,126 @@ export class AnomalyService extends MultiConnectionPoller implements OnModuleIni
...
+        ? ` The same address also hit other ACL denials (${otherReasons.join(', ')}), so it is ` +
+          `probing beyond the login itself.`
+        : '';
+    const usernameNote =
+      source.usernames.length > 1
+        ? ` across ${source.usernames.length} usernames (${users}${moreUsers}) — credential guessing rather than one stale client`
+        : ` against user '${users}'`;
+
+    return {
+      id: `${ctx.connectionId}-auth-failure-${source.clientAddress}-${timestamp}`,
```
[the remark this answers is not in the document]
Comment (jamby77) [reply proprietary/anomaly-detection/anomaly.service.ts]: Correct, and I owe you a correction on the sibling thread in #387 where I dismissed the same finding from CodeRabbit with confidently wrong evidence — my grep was silently failing on that file.

Verified properly this time: `CREATE TABLE IF NOT EXISTS anomaly_events (id UUID PRIMARY KEY, ...)` in `postgres.adapter.ts`. Fixed in d108f984 and across the stack.

The 10 remaining composite ids in that file are all pre-existing on `master`, so those detectors are memory-only on Postgres today. Separate issue, not this stack.

File: proprietary/anomaly-detection/auth-failure-detector.ts [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    string,
+    {
+      authFailures: number;
+      usernames: Map<string, number>;
+      reasonBreakdown: Record<string, number>;
+    }
+  >();
+
+  for (const { entry, windowCount } of accumulate(state, entries, now, windowMs)) {
+    const address = clientAddressFrom(entry.clientInfo);
```
[the remark this answers is not in the document]
Comment (jamby77) [reply proprietary/anomaly-detection/auth-failure-detector.ts:259]: Confirmed and fixed in 3ad9db5f — the root cause is in the store, not the detector.

`saveAclEntries` upserts on `(timestamp_created, username, object, reason, source_host, source_port, connection_id)` and its `DO UPDATE SET` refreshed `count`, `age_seconds`, `timestamp_last_updated` and `captured_at` — but not `client_info`. All three adapters had the same omission (SQLite, Postgres, and the equivalent spread in `MemoryAdapter`). So the address stayed pinned to whichever client happened to be recorded at first insert, which directly contradicts the detector doc, which describes the reported address as the most recent one.

Added `client_info` to the update in all three adapters and a new `acl-client-info-refresh.spec.ts` covering memory and SQLite. Verified the tests fail against the unpatched adapters (2 failed) and pass after (2 passed); the full `src/storage` suite is 173 passed / 17 suites, `tsc --noEmit` clean.

The detector itself is unchanged — with a truthful `client_info` its existing attribution is correct, and the "one entry can span several clients" caveat in the module doc still stands as written.

**Review by KIvanow — CHANGES_REQUESTED** [review]
1. The captured_at rationale is contradictory and wrong. The PR body's Scope note says the query "is windowed on captured_at" (and drops the ACL_DENIED-corroboration criterion on that basis), but the code does the opposite - no time filter, windowing on growth. And the code comment justifies that with "an entry under active attack keeps its original captured_at," yet the upsert sets captured_at = EXCLUDED.captured_at, so it's refreshed every save. The behavior is fine, but the description and the comment disagree with each other and with the upsert - please reconcile, since the scope decision rests on the (false) premise.

2. Multi-client attribution can name the wrong IP. ACL LOG groups by user+reason+object, not by client, and refreshing client_info to the latest means growth is attributed to whichever IP hit the entry last. Two attackers on one username merge into a single entry attributed to one IP, and attribution can flip between scans (per-address cooldown → repeat alerts naming different IPs). It's a documented ACL-LOG limitation and the message hedges well, but for an advisory that names an attacker IP it deserves prominent emphasis.

3. The client_info-on-upsert change is cross-cutting - it also changes what the ACL audit view shows for an existing entry (most-recent client, not first-seen), beyond this detector. Reasonable and tested, just worth calling out and confirming no audit-UI consumer relied on first-seen.

Nit: the 2000-cap is actually safe because ORDER BY captured_at DESC + refreshed captured_at keeps active entries on top - worth a comment saying so, since it reads as arbitrary right now.

**Comment by jamby77** [comment]
All four addressed in `ae86a947` plus a PR-body edit.

**1. `captured_at` contradiction — you were right, and the stated reason was simply false.** The upsert does `captured_at = EXCLUDED.captured_at`, so an entry under attack does *not* keep its original value. Reconciled all three places:

- **Code comment** now gives the real reason: `captured_at` records when the poller last *saw* the entry, not when the failures happened, and a restart or newly-added connection stamps every entry in the ring with the current time — so filtering on it would select on poller behaviour rather than attack activity. Growth-windowing is what excludes stale entries.
- **PR body** Scope note rewritten. The conclusion (no `ACL_DENIED` corroboration needed) survives, but it now rests on the true premise: a non-climbing entry contributes zero growth regardless of row age.

**4. The 2000-cap — folded into the same comment,** with your reasoning: because `captured_at` *is* refreshed and the store orders by `captured_at DESC`, entries still being written stay at the top and an active one cannot be pushed past the ceiling by dormant rows. That is exactly why the cap is safe, and it read as arbitrary before.

**2. Multi-client attribution — given the prominence you asked for.** The caveat is no longer a closing aside; it now leads with "read this as *the client last seen*, never a proven sole origin", and spells out both consequences explicitly: two attackers on one username merging into one entry attributed to whoever wrote last, and attribution flipping between scans so the per-address cooldown produces repeat advisories naming different IPs — not evidence of a second attack. Ends with "confirm against the raw `ACL LOG` before blocking an address."

**3. Cross-cutting `client_info` — checked, and nothing relied on first-seen.** Consumers are `AuditTrail.tsx`, which only displays the value and IP-filters on it, and `audit.service.ts`, which builds its messages from the freshly-parsed entry rather than from storage. For an audit trail, most-recent is arguably the more useful of the two. Called out here so it is on the record.

818 tests pass across anomaly-detection and storage.

**Review by KIvanow — APPROVED** [review]
